mono/metadata/threads.c

```diff
@@ -4095,13 +4096,33 @@ mono_threads_abort_appdomain_threads (MonoDomain *domain, int timeout)
 void
 ves_icall_thread_finish_async_abort (void)
 {
-	/* If we're in an abort protected block, just set the async requested bit and return.
-	 * Otherwise self abort now.
+	/* We were called from the handler block and are about to
+	 * leave it.  (If we end up postponing the abort because we're
+	 * in an abort protected block, the unwinder won't run and
+	 * won't clear the handler block itself which will confuse the
+	 * unwinder if we're in a try {} catch {} and we throw again.
+	 * ie, this:
+	 * static Constructor () {
+	 *   try {
+	 *     try {
+	 *       icall (); // Thread.Abort landed here,
+	 *                 // and caused the handler block to be installed
```

Comment:
@lambdageek  I think you meant for this icall () to be inside the finally clause ?

Reply:
I meant to show is that `icall` has a wrapper that has a `finally` clause. When we get the abort, we're in native and the handler block is installed on the `finally` of the wrapper.

Reply:
A handler block is installed only if we are inside a finally clause, not if we are guarded by a finally clause. Or am I still missing something from your explanation ?

Reply:
You're right.

This is what the victim thread looks like (abort-cctor.exe Test1) when the handler is installed:
```
#0  0xb7fdcd40 in __kernel_vsyscall ()
#1  0xb7f6a036 in sem_post@@GLIBC_2.1 ()
    at ../nptl/sysdeps/unix/sysv/linux/i386/i686/../i486/sem_post.S:62
#2  0x08322064 in mono_os_sem_post (sem=0x850b25c <suspend_semaphore>)
    at ../../mono/utils/mono-os-semaphore.h:282
#3  mono_threads_notify_initiator_of_suspend (info=0xb5400468) at mono-threads.c:106
#4  0x08326469 in mono_threads_state_poll_with_info (info=0xb5400468) at mono-threads-coop.c:147
#5  0x08175905 in mono_threads_safepoint () at ../../mono/utils/mono-threads-coop.h:43
#6  mini_method_compile (method=0x854ad40, opts=378629119, domain=0x8512bc0, flags=JIT_FLAG_RUN_CCTORS, 
    parts=0, aot_method_index=-1) at mini.c:3561
#7  0x08176175 in mono_jit_compile_method_inner (method=0x854ad40, target_domain=0x8512bc0, 
    opt=378629119, error=0xb56feb3c) at mini.c:4167
#8  0x08074751 in mono_jit_compile_method_with_opt (method=0x854ad40, opt=0, jit_only=0, error=0xb56feb3c)
    at mini-runtime.c:2200
#9  0x0810612b in common_call_trampoline (regs=0x1, regs@entry=0xb56feba0, 
    code=0x1 <error: Cannot access memory at address 0x1>, 
    code@entry=0xb5c807c0 <System_Threading_NativeEventCalls_SetEvent_Microsoft_Win32_SafeHandles_SafeWaitHandle+144> "\017\266E\354\205\300t\r\213E\b\211\004$9", m=0x854ad40, vt=0x0, vtable_slot=0x0, 
    error=0xb56feb3c) at mini-trampolines.c:739
#10 0x08106a7e in mono_magic_trampoline (regs=0xb56feba0, 
    code=0xb5c807c0 <System_Threading_NativeEventCalls_SetEvent_Microsoft_Win32_SafeHandles_SafeWaitHandle+144> "\017\266E\354\205\300t\r\213E\b\211\004$9", arg=0x854ad40, tramp=0x0) at mini-trampolines.c:872
#11 0xb7c15188 in ?? ()
#12 0x0854ad40 in ?? ()
#13 0xb5c6bc25 in System_Threading_EventWaitHandle_Set (this=...)
    at /home/aleksey/mono/mono/mcs/class/referencesource/mscorlib/system/threading/eventwaithandle.cs:319
#14 0xb5400f58 in ?? ()
#15 0xb7a10b20 in ?? ()
#16 0xb7a10930 in ?? ()
#17 0xb7a0c27d in ?? ()
#18 0x08075683 in mono_jit_runtime_invoke (method=0x85b82d8, obj=0x0, params=0x0, exc=0xb7401578, 
    error=0xb56fec78) at mini-runtime.c:2878
```

Frame 6 `method` arg is `"__icall_wrapper_mono_threads_state_poll"`, Frame 11 is `"<0xb7c15188 - generic_trampoline_jit trampoline>"` and Frame 12 stubbornly refuses to `mono_pmip`.  So I'm actually not sure where I am.